File: week3/lee/전력망을둘로나누기.py

def solution(n, wires): # 유니온 파인드로 풀었는데 왜 안되는지 모르겠음..
    def find(x):
        if(parents[x] != x):
            find(parents[x])
        return parents[x]
    
    def union(a, b):
        a = find(a)
        b = find(b)
        if a < b:
            parents[b] = a
        else:
            parents[a] = b

    answer = int(1e9)
    for idx, i in enumerate(wires):
        parents = [i for i in range(n+1)]
        for idx2,i2 in enumerate(wires):
            if(idx == idx2):
                continue
            else:
                union(i2[0], i2[1])
        sub1 = parents.count(find(i[0]))
        sub2 = parents.count(find(i[1]))
        answer = min(answer, abs(sub1-sub2))

    return answer

print(solution(4,[[1, 2], [1, 3], [1, 4]]	))

File: week3/eunbeen/전력망을둘로나누기.java
import java.util.*;

class Graph{
    int node;
    Set<Integer> list[];

    Graph(int node){
        this.node = node;
        this.list = new HashSet[node+1];

        for (int i=1; i<node+1; ++i){
            list[i] = new HashSet<Integer>();
        }
    }

    void edge(int a, int b){
        list[a].add(b);
        list[b].add(a);
    }

    void remove(int v1, int v2){
        list[v1].remove(v2);
        list[v2].remove(v1);
    }
    
    @Override 
    public String toString(){
        StringBuilder sb = new StringBuilder();
        for (int i=1; i<node+1; ++i){
            sb.append(i).append(": ");
            for(var item : list[i]){
                sb.append(item).append(" ");
            }
            sb.append("\n");
        }
        
       return sb.toString();
    }
}
class Solution {
    public int solution(int n, int[][] wires) {
        int answer = 10000;
        
        Graph graph = new Graph(n);
        
        for(int[] wire : wires){
            var v1 = wire[0];
            var v2 = wire[1];
            graph.edge(v1, v2);
        }
        
        for(int i=0; i<n-1; i++){
            var v1 = wires[i][0];
            var v2 = wires[i][1];
            
            graph.remove(v1, v2);
            int num1 = BFS(v1, graph, n);
            int num2 = BFS(v2, graph, n);
            
            answer = Math.min(answer, Math.abs(num1-num2));
            graph.edge(v1, v2);
        }
        
        return answer;
    }
    private int BFS(int index, Graph graph, int n){
        Queue<Integer> queue = new LinkedList();
        boolean[] check = new boolean[n+1];
        int count = 1;
        
        check[index] = true;
        queue.add(index);
        
        while(queue.size() > 0){
            index = queue.remove();
            
            Iterator<Integer> iter = graph.list[index].iterator();
            while(iter.hasNext()){
                int num = iter.next();
                if(!check[num]){
                    check[num] = true;
                    queue.add(num);
                    count += 1;
                }
            }
        }
        
        return count;
    }
}

// 전선을 하나씩 다 잘라서 확인해보기
// 1. 전선 하나 자르기
// 2. 한 공간의 개수 세기
// 3. 또 다른 공간의 개수 세기
// 4. 두 공간의 개수 차이 절대값 (min) 으로 구하기
// 5. 다시 전선 붙이기